src/DataLayouts/struct.jl

```diff
@@ -11,46 +11,92 @@
+        isbitstype(S) || error("Element type $S has indeterminate size")
+        is_valid_basetype(T, S) ||
+            error("Element type $S cannot be stored in container type $T")
+        return nothing
     end
-    FT1 = basetype(S1)
-    FT2 = basetype(Sx...)
-    FT1 !== FT2 && error("Inconsistent basetypes $FT1 and $FT2")
-    return FT1
 end
 
-replace_basetype(::Type{S}, ::Type{FT}) where {S <: AbstractFloat, FT} = FT
-function replace_basetype(::Type{S}, ::Type{FT}) where {S <: Tuple, FT}
-    Tuple{ntuple(i -> replace_basetype(fieldtype(S, i), FT), fieldcount(S))...}
-end
-function replace_basetype(
-    ::Type{NamedTuple{names, T}},
-    ::Type{FT},
-) where {names, T, FT}
-    NamedTuple{names, replace_basetype(T, FT)}
-end
+"""
+    replace_basetype(::Type{T}, ::Type{T′}, ::Type{S})
+
+Changes the type parameters of `S` to produce a new type `S′` such that, if
+`is_valid_basetype(T, S)` is true, then so is `is_valid_basetype(T′, S′)`.
+"""
+replace_basetype(::Type{T}, ::Type{T′}, ::Type{S}) where {T, T′, S} =
+    length(S.parameters) == 0 ? S :
+    S.name.wrapper{replace_basetypes(T, T′, S.parameters...)...}
+replace_basetype(::Type{T}, ::Type{T′}, ::Type{<:T}) where {T, T′} = T′
+replace_basetype(::Type{T}, ::Type{T′}, value) where {T, T′} = value
+replace_basetypes(::Type{T}, ::Type{T′}, value) where {T, T′} =
+    (replace_basetype(T, T′, value),)
+replace_basetypes(::Type{T}, ::Type{T′}, value, values...) where {T, T′} =
+    (replace_basetype(T, T′, value), replace_basetypes(T, T′, values...)...)
+# TODO: This could potentially lead to some annoying bugs, since it replaces
+# type parameters instead of field types. So, if `S` has `Float64` as a
+# parameter, `replace_basetype(Float64, Float32, S)` will replace that parameter
+# with `Float32`, regardless of whether the parameter corresponds to any values
+# stored in an object of type `S`. In other words, if a user utilizes types as
+# type parameters and specializes their code on those parameters, this may
+# change the results of their code.
+# Note that there is no way to write `replace_basetype` using field types
+# instead of type parameters, since replacing the field types of an object will
+# change that object's type parameters, but there is no general way to map field
+# types to type parameters.
 
 """
     parent_array_type(::Type{<:AbstractArray})
 
 Returns the parent array type underlying the `SubArray` wrapper type
 """
-parent_array_type(::Type{A}) where {A <: AbstractArray{FT}} where {FT} =
-    Array{FT}
-# TODO: extract interface to overload for backends into separate file
+parent_array_type(::Type{<:AbstractArray{T}}) where {T} = Array{T}
+
+"""
+    promote_container_type(::Type{<:AbstractArray}, ::Type{<:AbstractArray})
+
+Returns the type of a container that could store the same values as both input
+container types.
+"""
+promote_container_type(
+    ::Type{<:AbstractArray{T1}},
```

Comment:
I think this should just be enforced to be `Array{T1}` for now.  Like CUArrays we also need to enforce that the Dimension parameter is also correct ` where {T1, T2, Dim}`